core/editor/src/communication/dispatcher.rs

```diff
@@ -39,9 +40,11 @@ impl Dispatcher {
 			InputPreprocessor(message) => self.input_preprocessor.process_action(message, (), &mut self.messages),
 			InputMapper(message) => self.input_mapper.process_action(message, &self.input_preprocessor, &mut self.messages),
 		}
-		let message = self.messages.drain(..1).next();
-		if let Some(message) = message {
-			self.handle_message(message)?;
+		if !self.messages.is_empty() {
+			let message = self.messages.drain(..1).next();
+			if let Some(message) = message {
```

Comment:
Isn't this redundant? If `self.messages` isn't empty, then `drain(..1).next()` will always return `Some`, won't it?

Reply:
It panics, further cleanup required

Reply:
Sorry but I don't get what you're saying

Reply:
Drain panics if the vec is empty

Reply:
oh. If we're checking `is_empty` anyway then, we can replace `drain` with `remove(0)`

Reply:
or wait. This should have used `VecDeque` from the start